### PR TITLE
Resolve AJAX Select2 Issues

### PR DESCRIPTION
**Pull request by DanHarrin**
As #4413.

### CODE REVIEW

**Comment by bennothommo** [comment]
@DanHarrin Looks good. Have you tested it with both normal string-based keys as well as with all numeric keys too? My main concern with numeric keys is that the ordering will be off due to how browsers parse JSON objects.

**Comment by DanHarrin** [comment]
@bennothommo will have a look into that later, or maybe @LarBearrr can if he has time before me 😝

**Comment by w20k** [comment]
Re-compile assets, if that's the thing here 😉

**Comment by LarBearrr** [comment]
I made PR for the test plugin based on the old behavior. See [#75](https://github.com/octoberrain/test-plugin/pull/75).

**Comment by LarBearrr** [comment]
What do you mean @mjauvin ?

File: modules/system/assets/ui/js/select.js [hunk 1]
```diff
@@ -86,7 +86,23 @@
...
+                            if (results.hasOwnProperty(i)) {
+                                options.push({
+                                    id: i,
+                                    text: results[i],
+                                });
+                            };
+                        };
+
+                        return {
+                            results: options,
```
Comment (mjauvin) [inline modules/system/assets/ui/js/select.js]: here, I think key should be "result"

Comment (LarBearrr) [reply modules/system/assets/ui/js/select.js]: Ah, well, I'm basing this off of the [select2 docs for transforming data](https://select2.org/data-sources/ajax#transforming-response-data). 

I just tested with this code:
```
processResults: function (data,params) {
                        var results = data.result;
                        var options = [];

                        for (var i in results) {
                            if (results.hasOwnProperty(i)) {
                                options.push({
                                    id: i,
                                    text: results[i],
                                });
                            };
                        };                       
                        return {
                            results: options
                        };
```

And it is working. The only thing is that the items are not actually filtered based on your input. You just get a list of all options.

Comment (mjauvin) [reply modules/system/assets/ui/js/select.js]: ok, my bad.

Comment (LarBearrr) [reply modules/system/assets/ui/js/select.js]: No worries. Any thoughts on why the items are not filtered based on input? Is this something that has to happen in the backend?

**Comment by mjauvin** [comment]
The select2 doc you referred to says:

"Select2 expects results from the remote endpoint to be filtered on the server side."

**Comment by DanHarrin** [comment]
@LarBearrr that fix has been implemented

**Comment by DanHarrin** [comment]
@w20k recompiled the JS, we should be good to push

**Comment by w20k** [comment]
@mjauvin maybe you could run few tests? And we could proceed.

**Comment by bennothommo** [comment]
This pull request has been made to the wrong branch. Please review [the contributing guidelines](https://github.com/octobercms/october/blob/master/.github/CONTRIBUTING.md#pull-requests) as all PRs need to be made to the `develop` branch.

I'll fix it for you this time, but please ensure you make any future PRs to the `develop` branch, not the `master` branch.

**Comment by bennothommo** [comment]
@DanHarrin I've tested this and it works - as I expected, however, numeric keys are reordered by JavaScript when interpreting the JSON data received by the AJAX handler. This means, for example, if anyone uses this with an ID => name map, and have it ordered by name, all the results are out of order.

I think the best way to handle this is to have the `processResults()` method in the select.js allow for *both* types of data structure - ie. it should accept both:

```
$results = [
    'foo' => 'bar',
];
```
and
```
$results = [
    [
        'id' => 1,
        'text' => 'foo'
    ]
];
```
If we add both to the docs, it should allow for backwards compatibility as well as allowing the new format, and will give developers an informed choice on how to format the data.

**Comment by DanHarrin** [comment]
@bennothommo sorting it now 👍

**Comment by DanHarrin** [comment]
All done @bennothommo, please ensure that the fix I have implemented fits your criteria.

**Comment by w20k** [comment]
@bennothommo could you wait with merging, a little bit. Need to test few things.

**Comment by w20k** [comment]
@DanHarrin could you fix the docs and the check - should be isObject. Thought I was blind, but hunch told me something ain't right) JS array can't have an named id's. It's an object 😉

**Comment by DanHarrin** [comment]
@w20k the JS fix is done, but I'm unsure what on the docs needs fixing? Please guide me as to where they need to be updated.

File: modules/system/assets/ui/docs/select.md [hunk 2]
```diff
@@ -68,11 +68,28 @@ Use the `data-handler` attribute to source the select options from an AJAX handl
...
+></select>
 ```
 
 The AJAX handler should return results as an array.
 
+```php
+public function onGetOptions()
+{
+    $results = [
+        [
```
Comment (w20k) [inline modules/system/assets/ui/docs/select.md]: @DanHarrin  just change brackets here from `[` to `{` and `],` to `},`.

Comment (DanHarrin) [reply modules/system/assets/ui/docs/select.md]: @w20k but its PHP?

Comment (DanHarrin) [reply modules/system/assets/ui/docs/select.md]: @w20k the keyed array gets parsed to `{}` when it's passed to JS

Comment (w20k) [reply modules/system/assets/ui/docs/select.md]: @DanHarrin Sorry, still not waked up 😄 My bad, will provide you with the tips for JS in a minute)

**Comment by w20k** [comment]
@DanHarrin could you refactor this block, that's all from my side, @bennothommo (ping!):

```javascript
if (typeof i === 'object' && Array.isArray(i) === false) {
  options.push({
    id: results[i].id,
    text: results[i].text,
  });
} else {
   options.push({
      id: i,
      text: results[i],
   });
};
```
To a simple version
```
var isObject = i && i.constructor.name === "Object";
options.push({
  id: isObject ? results[i].id : i,
  text: isObject ? results[i].text : results[i],
});
```
JS note: typeof JavaScript constructors and objects (including null) returns "object"

**Comment by DanHarrin** [comment]
Done @w20k @bennothommo 👍

**Comment by LukeTowers** [comment]
@w20k is this good to go?

**Comment by w20k** [comment]
@LukeTowers yep ;)

**Comment by LukeTowers** [comment]
@w20k awesome, thanks!

**Pull request merged**